python/src/uagents/experimental/dialogues/__init__.py

```diff
@@ -0,0 +1,467 @@
+
+    def is_valid_message(self, session_id: UUID, msg_digest: str) -> bool:
+        """
+        Check if a message is valid for a given session.
+
+        Args:
+            session_id (UUID): The ID of the session to check the message for.
+            msg_digest (str): The digest of the message to check.
+
+        Returns:
+            bool: True if the message is valid, False otherwise.
+        """
+        if session_id not in self._sessions or len(self._sessions[session_id]) == 0:
+            return self.is_starter(msg_digest)
+
+        transition = self._resolve_mapping(msg_digest)
+        current_state = self.get_current_state(session_id)
+        current_state_mapped = self._resolve_mapping(current_state)
+        allowed_msgs = self._rules.get(current_state_mapped, [])
+        return transition in allowed_msgs
+
+    def is_valid_reply(self, in_msg: str, out_msg: str) -> bool:
+        """
+        Check if a reply is valid for a given message.
+
+        Args:
+            in_msg (str): The digest of the message to check the reply for.
+            out_msg (str): The digest of the reply to check.
+
+        Returns:
+            bool: True if the reply is valid, False otherwise.
+        """
+        return self._resolve_mapping(out_msg) in self._rules.get(
+            self._resolve_mapping(in_msg), []
+        )
+
+    def is_included(self, msg_digest: str) -> bool:
+        """
+        Check if a message is included in the dialogue.
+
+        Args:
+            msg_digest (str): The digest of the message to check.
+
+        Returns:
+            bool: True if the message is included, False otherwise.
+        """
+        return msg_digest in self._mapping.values()
+
+    def _load_storage(self) -> dict[UUID, list[Any]]:
+        """Load the sessions from the storage."""
+        cache: dict = self._storage.get(self.name)
+        return (
+            {UUID(session_id): session for session_id, session in cache.items()}
+            if cache
+            else {}
+        )
+
+    def _update_session_in_storage(self, session_id: UUID) -> None:
+        """Update a session in the storage."""
+        cache: dict = self._storage.get(self.name) or {}
+        cache[str(session_id)] = self._sessions[session_id]
+        self._storage.set(self.name, cache)
+
+    def _remove_session_from_storage(self, session_id: UUID) -> None:
+        """Remove a session from the storage."""
+        cache: dict = self._storage.get(self.name) or {}
+        session = str(session_id)
+        if session in cache.keys():
+            cache.pop(session)
+        self._storage.set(self.name, cache)
+
+    def _update_transition_model(self, edge: Edge, model: Type[Model]) -> None:
+        """Update the message model for a transition."""
+        self._mapping[edge.name] = Model.build_schema_digest(model)
+        self._edges[self._edges.index(edge)].model = model
+
+    def on_state_transition(self, edge_name: str, model: Type[Model]):
```

Comment:
I think there is a case for simply overriding `on_message` here and switching the order of `edge` and `model`.

If you do that, then the development pattern for adding a message handler is very similar to before, where the concept of `state_transition` is introduced as a required argument whenever you use the `Dialogue` class.

This also avoids having to disable `on_message` somehow.

But presumably you've already considered this, so maybe I am missing some disadvantage of doing this.